tools/cmdline.py

```diff
@@ -360,60 +353,65 @@ def consume_arg_file():
         # if we don't need to preserve LLVM debug info, do not keep this flag
         # for clang
-        if (settings.DEBUG_LEVEL < 3 and not
-            (settings.GENERATE_SOURCE_MAP or settings.SEPARATE_DWARF)):
+        if debug_level < 3 and not (settings.GENERATE_SOURCE_MAP or settings.SEPARATE_DWARF):
           newargs[i] = '-g0'
         else:
-          # for 3+, report -g3 to clang as -g4 etc. are not accepted
-          newargs[i] = '-g3'
-          if settings.DEBUG_LEVEL == 3:
+          if debug_level == 3:
             settings.GENERATE_DWARF = 1
-          if settings.DEBUG_LEVEL == 4:
-            settings.GENERATE_SOURCE_MAP = 1
+          elif debug_level == 4:
+            # In the past we supported, -g4.  But clang never did.
+            # Lower this to -g3, and report a warning.
+            newargs[i] = '-g3'
             diagnostics.warning('deprecated', 'please replace -g4 with -gsource-map')
+            settings.GENERATE_SOURCE_MAP = 1
+          elif debug_level > 4:
+            exit_with_error("unknown argument: '%s'", arg)
       else:
-        if requested_level.startswith('force_dwarf'):
+        if debug_level.startswith('force_dwarf'):
           exit_with_error('gforce_dwarf was a temporary option and is no longer necessary (use -g)')
-        elif requested_level.startswith('separate-dwarf'):
+        elif debug_level.startswith('separate-dwarf'):
           # emit full DWARF but also emit it in a file on the side
           newargs[i] = '-g'
           # if a file is provided, use that; otherwise use the default location
           # (note that we do not know the default location until all args have
           # been parsed, so just note True for now).
-          if requested_level != 'separate-dwarf':
-            if not requested_level.startswith('separate-dwarf=') or requested_level.count('=') != 1:
+          if debug_level != 'separate-dwarf':
+            if not debug_level.startswith('separate-dwarf=') or debug_level.count('=') != 1:
               exit_with_error('invalid -gseparate-dwarf=FILENAME notation')
-            settings.SEPARATE_DWARF = requested_level.split('=')[1]
+            settings.SEPARATE_DWARF = debug_level.split('=')[1]
           else:
             settings.SEPARATE_DWARF = True
           settings.GENERATE_DWARF = 1
           settings.DEBUG_LEVEL = 3
-        elif requested_level in ['source-map', 'source-map=inline']:
-          settings.GENERATE_SOURCE_MAP = 1 if requested_level == 'source-map' else 2
+        elif debug_level in ['source-map', 'source-map=inline']:
+          settings.GENERATE_SOURCE_MAP = 1 if debug_level == 'source-map' else 2
           newargs[i] = '-g'
-        elif requested_level == 'z':
+        elif debug_level == 'z':
           # Ignore `-gz`.  We don't support debug info compression.
-          continue
-        else:
+          pass
+        elif debug_level in ['line-tables-only', 'split-dwarf', 'dwarf-5']:
```

Comment:
Is there any way to figure out if this is comprehensive?

Reply:
Is this code run when linking too, or just compiling?
For compiling, clang has a bunch of debug flags that affect what goes into the DWARF, and that we don't want emcc to be concerned about, so the intention here was that we shouldn't be comprehensive (and the tradeoff is that maybe we miss the chance to diagnose a flag that we would have to care about but don't support currently).
But there's no existing precedent for -g flags when linking AFAIK, so I think at link time it would make sense to diagnose unrecognized flags. I'm not sure how feasible that is with the current implementation, and you could also make a case for ignoring unrecognized flags too.

Reply:
This code is run when both compiling and linking.

When compiling, adding an error here is redundant with clang which will already show the error.

When linking we currently do now show an error, as I mentioned above.   I'm not sure its worth fix though.

If we ran clang as our linker driver (and emld as the linker) which I have long wanted to do that would solve the issue I think.